File: java/Radix.java
package homework200323;

import java.util.Random;
import java.util.Timer;

public class Radix {
	
	static int getMaxNumber(int[] arr) {
		int max_num = arr[0];
		int arr_length = arr.length;
		for(int i = 0; i < arr_length; i ++) {
			if(arr[i] > max_num) {
				max_num = arr[i];
			}
		}
		return max_num;
	}
	
	static void countSort(int[] arr, int d) {
		int arr_length = arr.length;
		int[] final_arr = new int[arr_length];
		int[] count_arr = new int[10];
		
		for(int i = 0; i < 10; i++) {
			count_arr[i] = 0;
		}
		
		for(int i = 0; i < arr_length; i ++) {
			count_arr[(arr[i]/d)%10]++;
		}
		
		for(int i = 1; i < 10; i++) {
			count_arr[i] += count_arr[i-1];
		}
		
		for(int i = arr_length - 1; i >= 0; i--) {
			final_arr[count_arr[(arr[i]/d)%10] - 1] = arr[i];
			count_arr[(arr[i]/d)%10]--;
		}
		for(int i = 0; i < arr_length; i++) {
			arr[i] = final_arr[i];
		}
	}
	
	static void radixSort(int[] arr) {
		int max_numb = getMaxNumber(arr);
		
		for(int d = 1; max_numb/d > 0; d*=10) {
			countSort(arr,d);
		}
	}
	
	public static void main(String[] args) {
		int[] arr = new int[1000000];
		int arr_length = arr.length;
		Random r = new Random();
		for(int i = 0; i < arr_length; i++) {
			arr[i] = r.nextInt(100000);
		}
		long start = System.currentTimeMillis();
		radixSort(arr);
		long stop = System.currentTimeMillis();
		System.out.println("Radix time : " +  (stop-start));
	}
}

File: python/radix.py
import random
import timeit

arr = [0] * 1000000

for i in range(len(arr)):
  arr[i] = random.randrange(1,1000000)

def countingSort(arr, exp1): 
  
    n = len(arr) 
  
    output = [0] * (n) 
  
    count = [0] * (10) 
  
    for i in range(0, n): 
        index = (arr[i]//exp1) 
        count[ (index)%10 ] += 1
  
    # Change count[i] so that count[i] now contains actual 
    #  position of this digit in output array 
    for i in range(1,10): 
        count[i] += count[i-1] 
  
    # Build the output array 
    i = n-1
    while i>=0: 
        index = (arr[i]//exp1) 
        output[ count[ (index)%10 ] - 1] = arr[i] 
        count[ (index)%10 ] -= 1
        i -= 1
  
    i = 0
    for i in range(0,len(arr)): 
        arr[i] = output[i] 
  
def radixSort(arr): 
  
    max1 = max(arr) 
  
    exp = 1
    while max1/exp > 0: 
        countingSort(arr,exp) 
        exp *= 10

start = timeit.default_timer()
radixSort(arr)
stop = timeit.default_timer()
print("Radix Time", stop - start)